include/common/core/freertos_message_queue.hpp

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include <array>
+
+#include "FreeRTOS.h"
+#include "queue.h"
+
+namespace freertos_message_queue {
+
+template <typename Message, size_t queue_size = 10>
+class FreeRTOSMessageQueue {
+  public:
+    explicit FreeRTOSMessageQueue(const char* name) : FreeRTOSMessageQueue() {
+        vQueueAddToRegistry(queue, name);
+    }
+    explicit FreeRTOSMessageQueue()
+        : queue_control_structure(),
+          queue_data_structure(),
+          queue(xQueueCreateStatic(queue_size, sizeof(Message),
+                                   queue_data_structure.data(),
+                                   &queue_control_structure)) {}
+    FreeRTOSMessageQueue& operator=(FreeRTOSMessageQueue&) = delete;
+    FreeRTOSMessageQueue&& operator=(FreeRTOSMessageQueue&&) = delete;
+    FreeRTOSMessageQueue(FreeRTOSMessageQueue&) = delete;
+    FreeRTOSMessageQueue(FreeRTOSMessageQueue&&) = delete;
+
+    ~FreeRTOSMessageQueue() {
+        vQueueUnregisterQueue(queue);
+        vQueueDelete(queue);
+    }
+
+    auto try_write(const Message& message, const uint32_t timeout_ticks = 0)
+        -> bool {
+        return xQueueSendToBack(queue, &message, timeout_ticks) == pdTRUE;
+    }
+
+    auto try_read(Message* message, uint32_t timeout_ticks = 0) -> bool {
+        return xQueueReceive(queue, message, timeout_ticks);
+    }
+
+    [[nodiscard]] auto try_write_isr(const Message& message) -> bool {
+        BaseType_t higher_woken = pdFALSE;
+        auto sent = xQueueSendFromISR(queue, &message, &higher_woken);
+        portYIELD_FROM_ISR(  // NOLINT(cppcoreguidelines-pro-type-cstyle-cast)
+            higher_woken);
+        return sent;
+    }
+
+    auto try_read_isr(Message* message) const -> bool {
+        BaseType_t higher_woken = pdFALSE;
+        return xQueueReceiveFromISR(queue, message, &higher_woken) == pdTRUE;
```

Comment:
see `try_write_isr`, might need to do another `portYIELD` depending on `higher_woken`. should be in the freertos docs.

Reply:
yep it seems like it does need to be called.